### PR TITLE
feat(ci): switch to pre-commit.ci for code quality

### PR DESCRIPTION
**Pull request by neersighted**
pre-commit.ci is faster, and has less moving parts than our current Github Actions workflow. It integrates with Github checks, it can auto-correct PRs if desired (turned off in this PR), and it will automatically generate PRs to keep our `pre-commit` hooks up to date.

Note that this PR won't actually show pre-commit.ci working until an organization admin enables it (additionally, the protected branch config will have to be updated).

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (neersighted) [inline .pre-commit-config.yaml:18]: I've added commented out hooks that I think could benefit the project, but that currently would cause a lot of churn out-of-scope of this initial PR.

File: .pre-commit-config.yaml
Comment (neersighted) [inline .pre-commit-config.yaml:30]: All the pre-commit hooks are updated in this PR; pre-commit.ci will keep them up-to-date in the future. The non-config changes in this PR are a result of our out-of-date hooks.

File: .pre-commit-config.yaml
Comment (neersighted) [inline .pre-commit-config.yaml:27]: Currently getting an odd bug where `isort` tries to add two blank lines after imports in `.pyi` files, contrary to both typeshed and black conventions. If I can (later) bisect this properly I'll open an issue on isort.

File: .pre-commit-config.yaml
Comment (neersighted) [inline .pre-commit-config.yaml:39]: These are a harder sell, but still are useful tools. I think the mypy project should be completed before either. Some of the non-PEP8 compliant names may be part of the exposed external API and therefore require at least a minor version change to be corrected.

File: pyproject.toml
Comment (neersighted) [inline pyproject.toml:74]: Many of the options specified here were cargo-culted and did not do anything, or were done poorly. This is cleaned up to be as simple as I can manage given my experience with the tooling.

File: pyproject.toml
Comment (neersighted) [inline pyproject.toml:87]: Exclude actually wasn't doing anything here -- modern black has a much better approach that ignores files even when pre-commit passes them directly.

File: pyproject.toml
Comment (neersighted) [inline pyproject.toml:82]: This avoids clobbering the stock skip-list.

File: pyproject.toml
Comment (neersighted) [inline pyproject.toml:110]: This will cause merge conflicts with some of the mypi work, but I think this is a much better place to keep the config (pyproject.toml is meant to be the one file to rule them all, after all), and makes it much more concise.

File: pyproject.toml
Comment (neersighted) [inline pyproject.toml:94]: I've added supported Python versions to all the code quality tools -- it's some duplication, but it's all in pyproject.toml and prevents environment-specific differences.

**Comment by neersighted** [comment]
Incidentally, it looks like pre-commit.ci is going to run against all future PR pulls. To avoid unwanted auto-fix commits as much as possible, I can create a reduced subset of this that just disables the auto-fix commits until the larger PR can be merged.

image

File: pyproject.toml
Comment (branchvincent) [inline pyproject.toml:87]: this was the default (now extended to [include `.ipynb`](https://github.com/psf/black/blob/5e191c29d4492d14f9090a924a77b3dc055443a6/src/black/const.py#L3)), maybe just remove?

Comment (neersighted) [reply pyproject.toml:87]: Good catch -- it's in the black docs so somehow I assumed it was needed.

File: .pre-commit-config.yaml
Comment (neersighted) [inline .pre-commit-config.yaml:33]: These are gone because `black` provides them for Python files. If they're desired for other file types they can easily be added back.

**Comment by neersighted** [comment]
Fixed the failing checks by making the old `code-quality` workflow a no-op.

File: .pre-commit-config.yaml
Comment (abn) [reply .pre-commit-config.yaml:27]: I believe this is an isort regression, there is a historic version where this issue was resolved to align with the black profile.

See: https://github.com/PyCQA/isort/issues/1284

File: .pre-commit-config.yaml
Comment (abn) [reply .pre-commit-config.yaml:39]: I'd honestly remove unused hooks for now, and add them as and when we enable them.

**Review by abn — CHANGES_REQUESTED** [review]
Changes look good to me. I'd request that we remove the commented out checks for now and enable them as and when we need them.

File: .pre-commit-config.yaml
Comment (branchvincent) [inline .pre-commit-config.yaml:34]: fyi, `flake8` has moved to GitHub https://github.com/PyCQA/flake8 (gitlab is now just a mirror)

Comment (neersighted) [reply .pre-commit-config.yaml:34]: Good catch!

**Comment by neersighted** [comment]
> 
> 
> Changes look good to me. I'd request that we remove the commented out checks for now and enable them as and when we need them.

Done!

File: tests/repositories/fixtures/pypi.org/json/clikit.json
Comment (abn) [inline tests/repositories/fixtures/pypi.org/json/clikit.json:3]: Is this one expected?

Comment (neersighted) [reply tests/repositories/fixtures/pypi.org/json/clikit.json:3]: This is expected given https://simplejson.readthedocs.io/en/latest/#simplejson.dumps. There is no provision to disable this upstream -- I could add a argument if it's desired.

On the other hand, the spec allows for either encoded or non-encoded UTF-8 characters. This file is going to decode the same. If it's a bit too far (vs just re-indenting), I can roll it back until we get the ability to disable these escapes.

File: .pre-commit-config.yaml
Comment (neersighted) [reply .pre-commit-config.yaml:29]: Yeah, came across that as well. Pinged the maintainer -- for now I'm just going to leave the FIXME.

File: tests/repositories/fixtures/pypi.org/json/clikit.json
Comment (neersighted) [reply tests/repositories/fixtures/pypi.org/json/clikit.json]: It turns out I was looking at an old version of the hook and upstream already has support for turning this off.

https://github.com/pre-commit/pre-commit-hooks/blob/v4.0.1/pre_commit_hooks/pretty_format_json.py#L77

**Review by abn — APPROVED** [review]

**Pull request merged**
